### PR TITLE
Document Dependency Injection

### PR DESCRIPTION
**Pull request by jodosha**
Document how to use Dependency Injection with Granite.

#### Attributes

  * I want to avoid attributes to setup dependencies.
  * Mixing the two concepts is problematic as input (attributes) is a different concept than dependencies (e.g. a logger).

#### Specs

  * The specs are testing basic Ruby behavior.
  * But we’re NOT testing a basic Ruby object, we’re testing a Granite Action.
  * Granite Actions have attributes that handle getter and setters. I want to assert that it’s still possible to expose custom getters, without active_data to get on my way.
  * Those specs are trivial, but meant to prevent regressions: if we change attributes internals, we don’t accidentally override those behaviors.
  * These specs guarantee that documentation is up to date

### Review

- [x] Document code according to [Getting Started with Yard](http://www.rubydoc.info/gems/yard/file/docs/GettingStarted.md).
- [x] All tests are passing.
- [x] Test manually.
- [ ] Get approval.

### Pre-merge checklist

- [x] The PR relates to a single subject with a clear title and description in grammatically correct, complete sentences.
- [x] Verify that feature branch is up-to-date with `master` (if not - rebase it).
- [x] Double check the quality of [commit messages](http://chris.beams.io/posts/git-commit/).
- [x] Squash related commits together.

### CODE REVIEW

File: docs/index.md [hunk 1]
```diff
@@ -367,6 +367,27 @@ Adding errors to action object is important, because each time a handled excepti
 Validation exception will have the same backtrace as original error.
 Prefer this way over custom exception handling in private methods.
 
+### Dependency Injection
+
+Granite uses action initializer for attributes.
```
Comment (andriusch) [inline docs/index.md:372]: I don't really understand what you mean here? What is "action initializer"?

File: docs/index.md [hunk 2]
```diff
@@ -367,6 +367,27 @@ Adding errors to action object is important, because each time a handled excepti
...
+However, we can still manually inject dependencies.
```
Comment (andriusch) [inline docs/index.md:373]: I don't think we should be saying that this is how you should be doing dependency injection. I'm pretty sure that something like:
```ruby
Action.attribute :my_dep, Object
Action.new(my_dep: Bar.new)
```
Would also work. While I do see value in documenting ability to use custom initializers, something like: "here's how you can use custom parameters in cases where you don't want your data to end up in `action.attributes` (for example when using dependency injection)".

File: docs/index.md [hunk 3]
```diff
@@ -367,6 +367,27 @@ Adding errors to action object is important, because each time a handled excepti
...
+
+```ruby
+class Action < Granite::Action
+  attribute :name, String
+
+  private attr_reader :my_dep
```
Comment (andriusch) [inline docs/index.md:379]: Optional, but I think the example could be minified by removing this line since it's just basic ruby and not needed for the purpose of this example.

File: spec/lib/granite/action_spec.rb [hunk 4]
```diff
@@ -7,6 +7,55 @@
...
+          super(*args, **kwargs, &block)
+        end
+      end
+
+      stub_class(:another_action, Granite::Action) do
+        allow_if { true }
+      end
+    end
+
+    it 'creates private getter' do
```
Comment (andriusch) [inline spec/lib/granite/action_spec.rb:32]: This example is testing Ruby itself.

File: spec/lib/granite/action_spec.rb [hunk 5]
```diff
@@ -7,6 +7,55 @@
...
+      action = Action.new(comment: 'blah blah blah')
+      expect { action.my_dep }.to raise_error(NoMethodError, /private method `my_dep' called/)
+    end
+
+    it 'uses default value' do
```
Comment (andriusch) [inline spec/lib/granite/action_spec.rb:37]: Again you're testing ruby

File: spec/lib/granite/action_spec.rb [hunk 6]
```diff
@@ -7,6 +7,55 @@
     end
   end
 
+  describe 'Dependency Injection' do
```
Comment (andriusch) [inline spec/lib/granite/action_spec.rb:10]: I added a couple comments to individual tests, but tbh I really don't see value in any of these tests. You're testing basic ruby method overloading in subclass, nothing here seems granite specific.

**Review by andriusch — CHANGES_REQUESTED** [review]

File: docs/index.md (hunk 2 above)
Comment (jodosha) [reply docs/index.md:373]: @andriusch You're mixing up attributes with dependencies. Which is a big problem with current Granite implementation.

File: spec/lib/granite/action_spec.rb (hunk 6 above)
Comment (jodosha) [reply spec/lib/granite/action_spec.rb:10]: @andriusch This is to prevent regressions.

I want to ensure that the documentation is aligned with the code.
I want to ensure that if we change attributes internals, the documented code is still working and we don't accidentally override this behavior.

File: docs/index.md (hunk 1 above)
Comment (jodosha) [reply docs/index.md:372]: @andriusch `MyBa#initialize`.

File: spec/lib/granite/action_spec.rb (hunk 5 above)
Comment (jodosha) [reply spec/lib/granite/action_spec.rb:37]: @andriusch I'm testing that attributes would NOT interfere with the basics of Ruby.

File: docs/index.md (hunk 1 above)
Comment (andriusch) [reply docs/index.md:372]: Hm maybe reword this to something like "Granite assigns attributes in `initialize` method based on parameters that are passed to the initializer".

File: docs/index.md (hunk 2 above)
Comment (andriusch) [reply docs/index.md:373]: Sorry I think I didn't explain myself properly. I didn't mean that using `attributes` for DI is the correct approach, merely that it's a possible approach. My main concern here is that you're documenting only in regards to DI, but custom initializers could be used for other things as well, not only DI. So I'd prefer if it was worded something like: "You can still use custom initializers for Granite actions when you don't want to use attributes, for example when using dependency injection:". That way we're still documenting the preferred way to use DI, but also make it clear that this is not limited to DI.

File: spec/lib/granite/action_spec.rb (hunk 6 above)
Comment (andriusch) [reply spec/lib/granite/action_spec.rb:10]: But we can't override it outside of very aggressive shenanigans. Any subclass in Ruby can override `initialize` method and redefine what it does and define `attr_reader`s. Any method in Ruby can set instance variables.

The only 2 ways these tests could break due to changes in granite is:
- if we redefine what `attr_reader` does in Granite::Action.
- if we refefine what `private` does in Granite::Action.

And none of those sounds like something that would be allowed to merge.

File: docs/index.md (hunk 2 above)
Comment (jodosha) [reply docs/index.md]: Fixed in https://github.com/toptal/granite/pull/64/commits/57011bdcd89d18e1df00c9413582937c4cddc941

File: spec/lib/granite/action_spec.rb (hunk 5 above)
Comment (jodosha) [reply spec/lib/granite/action_spec.rb]: Removed by https://github.com/toptal/granite/pull/64/commits/2a98b4881927be094aac10d5ec3d882105bdba3d

File: spec/lib/granite/action_spec.rb (hunk 6 above)
Comment (jodosha) [reply spec/lib/granite/action_spec.rb:10]: @andriusch I'm more interested to prevent `active_data` regressions than `granite` ones.

That means, the tests are to protect us `active_data` changing the way attributes are handled. This isn't in `granite` as you're pointing. 🙂 

I agree with you that we won't do any of those changes in `granite`.

Given, I already removed [some controversial specs](https://github.com/toptal/granite/pull/64/commits/014e1e20cdf896f9b7d5c2eeb5e8f9cdb3d661b3), what's your take on this? To remove this `describe` block?

Comment (andriusch) [reply spec/lib/granite/action_spec.rb:10]: Yeah I'd suggest removing the whole describe. If your concern is `active_data`, I'm pretty sure we already have specs covering this, e.g. https://github.com/toptal/granite/blob/master/spec/lib/granite/action/subject_spec.rb#L32. `active_data` itself seems to be under maintenance mode so it's unlikely that they'll have any breaking API changes. TBH I think something needs to be done regarding `active_data`, but that's another story.

Comment (jodosha) [reply spec/lib/granite/action_spec.rb]: Removed https://github.com/toptal/granite/pull/64/commits/281a985a8cd2ccb32d9e4ed5305e3f97baf4aa9e

**Review by andriusch — APPROVED** [review]
